### PR TITLE
Playlist support & TV Shuffle

### LINKED ISSUE #108: FR: Playlist Support
Playlists are supported by the webapp and other clients, but return unknown media type in jellyfin-roku.

### PR DESCRIPTION
**Pull request by 1hitsong**
**Changes**
Adds support for Playlists from the home view. Currently it only supports playback for videos and audio.

It also adds a Shuffle button to TV series and seasons as a demo of how we will eventually use global queue for all playback.

This PR is an MVP product, so it is NOT feature complete.

This is currently a WIP, so there will be bugs. Please post any findings as a comment.

Ignore how much code duplication there is. I had to convert a ton of stuff to tasks to make this work. Eventually, if approved, all playback will use this system and the old code will be removed.

**Issues**
Fixes #108 #198

### CODE REVIEW

**Review by candry7731 — APPROVED** [review]
This is Great! I used the playlist with mixed content and it worked flawlessly. It will be nice to have more features like, shuffle and random

The Shuffle and random buttons on on tv shows is a great addition! Also worked on my server with no bugs.

It would be nice if the shuffle and play random buttons were the same size. 
image

File: components/music/AudioPlayerView.brs [hunk 1]
```diff
@@ -227,7 +229,9 @@ sub audioStateChanged()
 
         if m.global.queueManager.callFunc("getPosition") < m.global.queueManager.callFunc("getCount") - 1
             ' We are not at the end of the song queue, advance to next song
-            LoadNextSong()
+            'LoadNextSong()
```
Comment (jimdogx) [inline components/music/AudioPlayerView.brs]: Remove commented out code.

File: components/video/VideoPlayerView.brs [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    '    m.top.control = "stop"
+    '    return true
+    'end if
+
+    'if m.previousItem.hasFocus() and key = "OK"
+    '    m.global.queueManager.callFunc("moveBack")
+    '    m.global.queueManager.callFunc("playQueue")
+    '    m.top.control = "stop"
+    '    return true
+    'end if
```
Comment (jimdogx) [inline components/video/VideoPlayerView.brs]: Commented out code.

File: components/ItemGrid/LoadVideoContentTask.brs [hunk 3]
```diff
@@ -0,0 +1,1258 @@
...
+
+    resumeData = [
+        "Resume playing at " + ticksToHuman(time) + ".",
+        "Start over from the beginning."
+    ]
+
+    m.global.sceneManager.callFunc("optionDialog", tr("Playback Options"), ["Choose an option"], resumeData)
+
+    while not isValid(m.global.sceneManager.returnData)
+
```
Comment (jimdogx) [inline components/ItemGrid/LoadVideoContentTask.brs:257]: This looks scary, but maybe it's not an issue.  Seems like there should be some sort of timeout that eventually throws an error.

File: components/ItemGrid/LoadVideoContentTask.brs [hunk 4]
```diff
@@ -0,0 +1,1258 @@
...
+    if m.user.Configuration.EnableNextEpisodeAutoPlay
+        ' query API for next episode ID
+        url = Substitute("Shows/{0}/Episodes", showID)
+        urlParams = { "UserId": get_setting("active_user") }
+        urlParams.Append({ "StartItemId": videoID })
+        urlParams.Append({ "Limit": 2 })
+        resp = APIRequest(url, urlParams)
+        data = getJson(resp)
+
+        if data <> invalid and data.Items.Count() = 2
```
Comment (jimdogx) [inline components/ItemGrid/LoadVideoContentTask.brs:336]: Should we be standardizing on isValid() throughout?

**Review by jimdogx — APPROVED** [review]
Approved with some "cleanup" nitpicks.

File: components/ItemGrid/LoadVideoContentTask.brs (hunk 3 above)
Comment (1hitsong) [reply components/ItemGrid/LoadVideoContentTask.brs:257]: The while is simply waiting for user input. 

I suppose we could implement a timeout. If the user hasn't made a selection within X seconds, we close the popup for them. What do you think?

File: components/video/VideoPlayerView.brs (hunk 2 above)
Comment (1hitsong) [reply components/video/VideoPlayerView.brs]: I'll remove it to keep the PR clean. This is my first code supporting playlist navigation - allowing the user to move back and forward within the queue.

File: components/ItemGrid/LoadVideoContentTask.brs (hunk 4 above)
Comment (1hitsong) [reply components/ItemGrid/LoadVideoContentTask.brs:336]: Eventually, I want to use isValid() throughout, but most of that code in that task was copy/paste from other spots with only minor updates to make it work with a task. I tried to keep the changes there to only essentials.

But yeah, that'll be part of the cleanup as it gets refined.

File: components/ItemGrid/LoadVideoContentTask.brs (hunk 3 above)
Comment (jimdogx) [reply components/ItemGrid/LoadVideoContentTask.brs:257]: For some reason I missed that this was calling "optionDialog"... my eyes just focused on this never ending while loop haa haa.

**Comment by 1hitsong** [comment]
@jkim2492 This will impact your work on the subtitles. The end goal is to unify all playback into the global queue functions. To make this work, I had to duplicate a lot of the video playback code and modify it to work with tasks.

This PR addresses playlist support and shuffle tv seasons. I'm going to start applying it to the other video playback calls, but before I do, I need to copy over your new subtitle functions into the new video player.

Before you continue work on the embeded subtitles, you may want to either wait until I finish this, or work on adding it here to begin with.

**Comment by jkim2492** [comment]
@1hitsong  I think waiting won't be a problem since I haven't decided on how to implement it yet :)

File: components/ItemGrid/LoadVideoContentTask.brs [hunk 5]
```diff
@@ -0,0 +1,1258 @@
...
+                        return false
+                    end if
+                end if
+            end while
+        end if
+    end if
+    return true
+end function
+
+function CreateMovieDetailsGroup(movie)
```
Comment (cewert) [inline components/ItemGrid/LoadVideoContentTask.brs:1196]: im fixing the crashlog error for ShowScenes.brs and noticed the `CreateMovieDetailsGroup` function was duplicated here. can't we just include ShowScenes.brs to this task node?

**Pull request merged**